### PR TITLE
planner: refactor the function 

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #36598 

Problem Summary:
The `GetPlanFromSessionPlanCache` function has too much logic to read，before general plan cache we need do some finishing.

### What is changed and how it works?
Split it into 4 functions.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache.go [hunk 1]
```diff
@@ -70,10 +71,32 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
 		}
 	}
 
+	isBinProtocol := len(binProtoVars) > 0
+	varsNum, binVarTypes, txtVarTypes := getParamValue(sctx, isBinProtocol, binProtoVars, txtProtoVars)
+
+	if prepared.UseCache && prepared.CachedPlan != nil && !ignorePlanCache { // for point query plan
```
Comment (dayicklp) [inline planner/core/plan_cache.go:77]:
```suggestion
	if prepared.UseCache && prepared.CachedPlan != nil && !ignorePlanCache { // For the point query plan, we need ....
```

File: planner/core/plan_cache.go [hunk 2]
```diff
@@ -70,10 +71,32 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+		if plan, names, done, err := getPointQueryPlan(prepared, sessVars, stmtCtx); done == true {
```
Comment (dayicklp) [inline planner/core/plan_cache.go:78]:
```suggestion
		if plan, names, done, err := getPointQueryPlan(prepared, sessVars, stmtCtx) && done {
```

Comment (dayicklp) [reply planner/core/plan_cache.go:78]: I'm not sure that's what it means

**Review by hawkingrei — APPROVED** [review]

File: build/nogo_config.json [hunk 3]
```diff
@@ -253,6 +253,7 @@
       "planner/util": "planner code",
       "planner/implementation": "planner code",
       "planner/cascades": "planner code",
+      "planner/core/plan_cache.go": "planner code",
```
Comment (qw4990) [inline build/nogo_config.json:256]: Should we remove this before merging this PR?

Comment (hawkingrei) [reply build/nogo_config.json:256]: No, let's not make this file break again.

File: planner/core/plan_cache.go [hunk 4]
```diff
@@ -70,10 +71,32 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+		}
+	}
+
+	return rebuildPhysicalPlan(ctx, sctx, is, preparedStmt, ignorePlanCache, cacheKey,
+		latestSchemaVersion, isBinProtocol, varsNum, binVarTypes, txtVarTypes)
+}
+
+// getParamValue get real values for "?" in PREPARE statements
+func getParamValue(sctx sessionctx.Context, isBinProtocol bool, binProtoVars []types.Datum,
+	txtProtoVars []expression.Expression) (int, []byte, []*types.FieldType) {
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:96]: Maybe we can use 
```suggestion
	txtProtoVars []expression.Expression) (varsNum int, binVarTypes []byte, txtVarTypes []*types.FieldType) {
```

File: planner/core/plan_cache.go [hunk 5]
```diff
@@ -90,90 +113,105 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
-		if metrics.ResettablePlanCacheCounterFortTest {
-			metrics.PlanCacheCounter.WithLabelValues("prepare").Inc()
-		} else {
-			planCacheCounter.Inc()
-		}
-		sessVars.FoundInPlanCache = true
-		stmtCtx.PointExec = true
-		return plan, names, nil
+func getPointQueryPlan(prepared *ast.Prepared, sessVars *variable.SessionVars, stmtCtx *stmtctx.StatementContext) (plan Plan,
+	names []*types.FieldName, done bool, err error) {
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:120]: Maybe rename `done` to `ok` is more common in the golang.

Comment (fzzf678) [reply planner/core/plan_cache.go:120]: Get

File: planner/core/plan_cache.go [hunk 6]
```diff
@@ -70,10 +71,32 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+	return rebuildPhysicalPlan(ctx, sctx, is, preparedStmt, ignorePlanCache, cacheKey,
+		latestSchemaVersion, isBinProtocol, varsNum, binVarTypes, txtVarTypes)
+}
+
+// getParamValue get real values for "?" in PREPARE statements
+func getParamValue(sctx sessionctx.Context, isBinProtocol bool, binProtoVars []types.Datum,
+	txtProtoVars []expression.Expression) (varsNum int, binVarTypes []byte, txtVarTypes []*types.FieldType) {
 	var varsNum int
 	var binVarTypes []byte
 	var txtVarTypes []*types.FieldType
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:99]: This part can be removed. And we can just `return` in the end.

File: planner/core/plan_cache.go [hunk 7]
```diff
@@ -90,90 +113,105 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+		metrics.PlanCacheCounter.WithLabelValues("prepare").Inc()
+	} else {
+		planCacheCounter.Inc()
+	}
+	sessVars.FoundInPlanCache = true
+	stmtCtx.PointExec = true
+	return plan, names, true, nil
+}
+
+func getGeneralPlan(ctx context.Context, sctx sessionctx.Context, sessVars *variable.SessionVars, stmtCtx *stmtctx.StatementContext,
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:143]: I think the input parameters are so many. Maybe some input parameter can be removed. And generate in this function from other parameter.
For example, we can use the `ctx` and `sctx` to get the `sessVars` and `stmtCtx`.

File: planner/core/plan_cache.go [hunk 8]
```diff
@@ -90,90 +113,105 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
 
-REBUILD:
 	planCacheMissCounter.Inc()
 	stmt := prepared.Stmt
 	p, names, err := OptimizeAstNode(ctx, sctx, stmt, is)
 	if err != nil {
-		return nil, nil, err
+		return p, names, err
 	}
+	//err = e.tryCachePointPlan(ctx, sctx, preparedStmt, is, p)
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:211]: This line can be removed.

File: planner/core/plan_cache.go [hunk 9]
```diff
@@ -90,90 +113,105 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+	sessVars := sctx.GetSessionVars()
+	stmtCtx := sessVars.StmtCtx
 
-REBUILD:
 	planCacheMissCounter.Inc()
 	stmt := prepared.Stmt
 	p, names, err := OptimizeAstNode(ctx, sctx, stmt, is)
 	if err != nil {
-		return nil, nil, err
+		return p, names, err
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:209]: Why change here? The following are same.

Comment (fzzf678) [reply planner/core/plan_cache.go:209]: For Looks unified，i will change it back

Comment (fzzf678) [reply planner/core/plan_cache.go:209]: > Why change here? The following is same.

For Looks unified，i will change it back

File: planner/core/plan_cache.go [hunk 10]
```diff
@@ -90,81 +110,98 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+	// Rewriting the expression in the select.where condition  will convert its
+	// type from "paramMarker" to "Constant".When Point Select queries are executed,
+	// the expression in the where condition will not be evaluated,
+	// so you don't need to consider whether prepared.useCache is enabled.
+	plan := prepared.CachedPlan.(Plan)
+	names := prepared.CachedNames.(types.NameSlice)
+	err := RebuildPlan4CachedPlan(plan)
+	if err != nil {
+		logutil.BgLogger().Debug("rebuild range failed", zap.Error(err))
+		return plan, names, false, nil
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:128]:
```suggestion
		return nil, nil, false, nil
```

**Review by Reminiscent — APPROVED** [review]
Rest LGTM. @qw4990 please take a look

File: planner/core/plan_cache.go [hunk 11]
```diff
@@ -70,10 +71,29 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+			return plan, names, err
+		}
+	}
+
+	return rebuildPhysicalPlan(ctx, sctx, is, preparedStmt, ignorePlanCache, cacheKey,
+		latestSchemaVersion, isBinProtocol, varsNum, binVarTypes, txtVarTypes)
+}
+
+// getParamValue get real values for "?" in PREPARE statements
+func getParamValue(sctx sessionctx.Context, isBinProtocol bool, binProtoVars []types.Datum,
```
Comment (qw4990) [inline planner/core/plan_cache.go:95]: How about renaming it to `parseParamTypes`?

Comment (fzzf678) [reply planner/core/plan_cache.go:95]: > How about renaming it to `parseParamTypes`?

Oops，it's indeed `type` not `value`

File: planner/core/plan_cache.go [hunk 12]
```diff
@@ -90,81 +110,98 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
...
+				stmtCtx.SetPlanDigest(preparedStmt.NormalizedPlan, preparedStmt.PlanDigest)
+				return cachedVal.Plan, cachedVal.OutPutNames, true, nil
 			}
+			break
 		}
 	}
+	return nil, nil, false, nil
+}
+
+func rebuildPhysicalPlan(ctx context.Context, sctx sessionctx.Context, is infoschema.InfoSchema, preparedStmt *CachedPrepareStmt,
```
Comment (qw4990) [inline planner/core/plan_cache.go:198]: This function is used to generate a new plan for the current statement if there is no valid plan for it.
So a name like `generateNewPlan` seems better. Please rename it and add some comments.

**Review by qw4990 — APPROVED** [review]
LGTM

**Pull request merged**